### ISSUE TITLE
Memory leak in callbacksRegistry / setApplicationMenu

### ISSUE THREAD

**Issue by patrickhulce** [electron/electron#24187]
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:**
  * at least 6.0.12 up to latest 9.0.4
* **Operating System:**
  * macOS
* **Last Known Working Electron version:**
  * Unknown

### Expected Behavior
Subsequent calls to `setApplicationMenu` dereference the previous menu object so it and all its handlers can be garbage collected.

### Actual Behavior
All menu listeners (along with their closures) that are used in `setApplicationMenu` are retained forever with no way to dereference them :(

I've tried a few workarounds like `Menu.setApplicationMenu(null)`

### To Reproduce
I've created a [mini repo](https://github.com/patrickhulce/electron-performance-examples/blob/1dec2b732bf5ea2baf9e78d3c03adb08df844877/index.html#L37-L51) with some performance examples that allow you to play around with this.

tl;dr - call setApplicationMenu a bunch of times with a large event listener
```js
for (let i = 0; i < 100; i++) {

const extraLargeObject = {}
for (let i = 0; i < 10000; i++) {
  extraLargeObject[`cool-key${i}`] = `super long string ${Date.now()}`.repeat(100)
}

const {Menu} = require('electron').remote
const menu = Menu.buildFromTemplate([
  {
    label: 'Menu Option',
    submenu: [{label: 'Action', click: () => console.log(extraLargeObject)}],
  },
])
Menu.setApplicationMenu(menu)

}
```

### Screenshots
heap snapshot showing all historical `click` handlers stored in callbacksRegistry
![image](https://user-images.githubusercontent.com/2301202/84960061-4ff9b700-b0c6-11ea-8bb1-74d418e2e0f6.png)

### Notes
I would also be satisfied with a workaround for how to ensure that the previous menu is cleaned up in the renderer :) 

We discovered this in our application that runs for a very long time which always eventually OOMs due to this problem. The only workaround I see is to ensure that you only ever invoke `setApplicationMenu` exactly once with listeners that simply delegate to the appropriate template items.

**Comment by nornagon** [comment]
I'd recommend [not using the remote module](https://medium.com/@nornagon/electrons-remote-module-considered-harmful-70d69500f31).

**Comment by patrickhulce** [comment]
Thanks @nornagon! We actually already started moving off remote already after that article and have fixed this problem in our app by moving to custom IPC flow for this 👍  

Just thought this was something still worth bringing up. Being permanently unable to clear the callbacks in `callbacksRegistry` was a monster user-facing app stability issue for us whereas the other demons you've already highlighted had a work around, were less user impactful, or just weren't a primary concern for the business.

**Comment by nornagon** [comment]
Yeah, this sort of issue is one of the major reasons we decided to deprecate the remote module in the first place. This is maybe fixable (assuming it doesn't involve reference cycles between the main and renderer process, which are not detectable), but tracking down this kind of issue is incredibly difficult.

I'm marking this WONTFIX as I don't plan to work on it, but PRs welcome if you do happen to track it down.
